File: inhouse/rc.cc
// 786
// benchmarks:
// - reverse complementing 10g FASTA

#include <iostream>
#include <fstream>
#include <vector>
using namespace std;

char revcomp(char c) {
   return (c == 'T' ? 'A' :
          (c == 'G' ? 'C' :
          (c == 'C' ? 'G' :
          (c == 'A' ? 'T' : c))));
}

int rc_copy(string s) {
   string x = s;
   for (int i = 0; i < x.size(); i++)
      x[i] = revcomp(s[s.size() - i - 1]);
   cout << x << endl;
   return x.size();
}

int main(int argc, char **argv) {
   ios_base::sync_with_stdio(false);
	 cin.tie(nullptr);

   ifstream fin(argv[1]);
   string s;
   long long total = 0;
   while (getline(fin, s)) {
      total += rc_copy(s);
   }
   cout << total << endl;

   return 0;
}

File: inhouse/rc.py
# 786
# benchmarks:
# - reverse complementing 10g FASTA
# - printing all 16-mers from 10g FASTA
# - count number of CpG islands (e.g. all CG-only subseqs)

import sys

def revcomp(c):
   return ('A' if c == 'T' else \
          ('C' if c == 'G' else \
          ('G' if c == 'C' else \
          ('T' if c == 'A' else c))))

def rc_copy(s):
   rc = ''.join(
      revcomp(s[len(s) - i - 1]) for i in range(len(s)))
   print rc
   return len(rc)

f = open(sys.argv[1], 'r')
total = 0
for l in f:
   total += rc_copy(l.strip())
print total
